cmd/multiFaExtract/multiFaExtract.go

```diff
@@ -5,54 +5,102 @@ package main
 import (
 	"flag"
 	"fmt"
+	"github.com/vertgenlab/gonomics/bed"
 	"github.com/vertgenlab/gonomics/common"
 	"github.com/vertgenlab/gonomics/fasta"
 	"log"
 )
 
-func multiFaExtract(infile string, outfile string, start int, end int, removeGaps bool) {
-	if !(start < end) {
-		log.Fatalf("Invalid arguments, start must be lower than end")
-	}
-	records := fasta.Read(infile)
+func multiFaExtract(s Settings) {
 	var ans []fasta.Fasta
-	for i := 0; i < len(records); i++ {
-		ans = append(ans, fasta.Extract(records[i], fasta.RefPosToAlnPos(records[0], start), fasta.RefPosToAlnPos(records[0], end), records[i].Name))
-	}
-	if removeGaps {
-		ans = fasta.RemoveGaps(ans)
+	records := fasta.Read(s.InFile)
+	if s.Bed == "" {
+		if !(s.Start < s.End) {
+			log.Fatalf("Invalid arguments, start must be lower than end")
+		}
+		ans = extractMultiHelper(records, s.Start, s.End)
+		if s.RemoveGaps {
+			ans = fasta.RemoveGaps(ans)
+		}
+		fasta.Write(s.OutFile, ans)
+	} else {
+		bedChan := bed.GoReadToChan(s.Bed)
+		for b := range bedChan {
+			ans = extractMultiHelper(records, b.ChromStart, b.ChromEnd)
+			if s.RemoveGaps {
+				ans = fasta.RemoveGaps(ans)
+			}
+			fasta.Write(fmt.Sprintf("%s.%v.%v.fa", b.Chrom, b.ChromStart, b.ChromEnd), ans)
+		}
 	}
+}
 
-	fasta.Write(outfile, ans)
+func extractMultiHelper(records []fasta.Fasta, start int, end int) []fasta.Fasta {
+	var ans = make([]fasta.Fasta, len(records))
+	for i := range records {
+		ans[i] = fasta.Extract(records[i], fasta.RefPosToAlnPos(records[0], start), fasta.RefPosToAlnPos(records[0], end), records[i].Name)
+	}
+	return ans
 }
 
 func usage() {
 	fmt.Print(
 		"multiFaExtract - Pull sub-sequence from multiple Fasta alignment for each entry. Uses reference indices, treating the first sequence as the reference.\n" +
 			"Usage:\n" +
 			"multiFaExtract multi.fa out.fa start end\n" +
+			"OR\n" +
+			"multiFaExtract -bed regions.bed multi.fa\n" +
 			"options:\n")
 	flag.PrintDefaults()
 }
 
+type Settings struct {
```

Comment:
That's cool. So if I'm understanding correctly, you took what used to be the original required input and put it into this struct so that you can use an simple list of "if" statements to determine if the user would like to use the original way or use your new way, which is with the bed input? And then by having a struct, you have a simple way of defining the difference between the 'original' versus the 'bed' way.

Reply:
yes exactly. As we have a bunch of related variables that all need to be passed around together it made sense to envelope it in a struct